app.py

```diff
@@ -39,12 +39,22 @@ def home():
 
 @app.route("/login")
 def login():
+    # store app state and code verifier in session
+    session['app_state'] = secrets.token_urlsafe(64)
+    session['code_verifier'] = secrets.token_urlsafe(64)
+
+    # calculate code challenge
+    hashed = hashlib.sha256(session['code_verifier'].encode('ascii')).digest()
+    encoded = base64.urlsafe_b64encode(hashed)
+    code_challenge = encoded.decode('ascii')[:-1]
```

Comment:
This makes me nervous, are you sure there will always be a single `=` at the end of the encoded string? It might be safer to use a function that trims the `=` off the end.

```suggestion
    code_challenge = encoded.decode('ascii').strip('=')
```